_posts/2021-03-29-apply-pretrained-pytorch-model.md

```diff
@@ -0,0 +1,154 @@
+3. Construct a function to apply the model onto each chunk
+4. Apply that function across the Dask array with the dask.array.map_blocks function.
+5. Store the result back into Zarr format
+
+### Step 1. Load the image data
+
+First, we load the image data into a Dask array.
+
+ We were provided sample data from [Prof. Shroff’s lab](https://www.nibib.nih.gov/about-nibib/staff/hari-shroff) at the NIH. The data originally was provided as a 3D TIFF file which we subsequently converted to Zarr with a shape of (950, 2048, 2048). That's a z-stack of 950 images where each slice is 2048x2048 pixels.
+
+This is the same data that we analysed in our last [blogpost on Dask and ITK](https://blog.dask.org/2019/08/09/image-itk). You should note the similarities to that workflow even though we are now using new libaries and performing different analyses.
+
+```
+cd '/Users/nicholassofroniew/Github/image-demos/data/LLSM'
+```
+
+```python
+# Load our data
+import dask.array as da
+imgs = da.from_zarr("AOLLSM_m4_560nm.zarr")
+imgs
+```
+
+```
+dask.array<from-zarr, shape=(20, 199, 768, 1024), dtype=float32, chunksize=(1, 1, 768, 1024)>
+```
+
+### Step 2. Load a pre-trained PyTorch model
+
+Next, we load our pre-trained UNet model.
+
+This UNet model takes in an 2D image and returns a 2D x 16 array, where each pixel is now associate with a feature vector of length 16.
+
+We thank Mars Huang for training this particular UNet on a corpous of biological images to produce biologically relevant feature vectors, as part of his work on [interactive bio-image segmentation](https://github.com/transformify-plugins/segmentify). These features can then be used for more downstream image processing tasks such as image segmentation.
+
+
+```python
+# Load our pretrained UNet¶
+import torch
+from segmentify.model import UNet, layers
+
+def load_unet(path):
+    """Load a pretrained UNet model."""
+
+    # load in saved model
+    pth = torch.load(path)
+    model_args = pth['model_args']
+    model_state = pth['model_state']
+    model = UNet(**model_args)
+    model.load_state_dict(model_state)
+
+    # remove last layer and activation
+    model.segment = layers.Identity()
+    model.activate = layers.Identity()
+    model.eval()
+
+    return model
+
+model = load_unet("HPA_3.pth")
+```
+
+### Step 3. Construct a function to apply the model to each chunk
+
+We make a function to apply our pre-trained UNet model to each chunk of the Dask array.
+
+Because Dask arrays are just made out of Numpy arrays which are easily converted to Torch arrays, we're able to leverage the power of machine learning at scale.
+
+```python
+# Apply UNet featurization
+import numpy as np
+
+def unet_featurize(image, model):
+    """Featurize pixels in an image using pretrained UNet model.
+    """
+    import numpy as np
+    import torch
+
+    # Extract the 2D image data from the Dask array
+    # Original Dask array dimensions were (time, z-slice, y, x)
```

Comment:
Is this correct @sofroniewn?

Reply:
yes, i should probably add a comment that notes that line 104 and line 108 are redundant, or maybe we drop both those lines and just have a comment. The unet expects 4D `(Batch x Channel x Width x Height)` the `Original Dask array dimensions were (time, z-slice, y, x)` it just so happens those match, but they need not have ...

Reply:
I *don't* want us to drop those lines from the code. For anyone who copy-pastes this example and then tries to tweak it to fit their data, these are important steps in the process (even if they do happen to be redundant in this particular example).

We could a line of text underneath making a small note of this point.

Reply:
I've added a note.